### PR TITLE
User Provided Args

### PR DESCRIPTION
**Pull request by BenKurrek**

### CODE REVIEW

File: contract/src/stage1/function_call.rs [hunk 1]
```diff
@@ -23,50 +25,66 @@ impl Keypom {
...
-            if final_args.contains("\"keypom_args\"") {
+            if final_args.len() == 0 {
+                final_args = "{}".to_string();
+            }
+            
+            let try_json: Result<Value, _> = from_str(&final_args);
+            if try_json.is_err() {
                 near_sdk::log!(
-                    "Keypom Args detected in client args. Returning and decrementing keys"
+                    "Cannot cast args to JSON. Returning and decrementing keys"
```
Comment (mattlockyer) [inline contract/src/stage1/function_call.rs:44]: Should we tell them decrementing keys effect? Like do we decrement uses, gas, both, etc...?

File: contract/src/stage1/function_call.rs [hunk 2]
```diff
@@ -23,50 +25,66 @@ impl Keypom {
...
+                            final_args_json = user_args_json;
+                        }
+                        UserArgsRule::UserPreferred => {
+                            // Take the final args string and merge the user args into it and overwrite any duplicate keys
+                            merge_json(&mut final_args_json, &user_args_json);
+                        }
+                    }
+                } else {
+                    near_sdk::log!(
+                        "Cannot cast user provided args to JSON. Disregarding user args and continuing"
```
Comment (mattlockyer) [inline contract/src/stage1/function_call.rs]: is this expected behavior? Seems like something we're choosing / should warn the user about?

I think we should do the same as above, nothing. IMHO.

{"myProposal":"something } in the wrong format"}

uh oh! proposal created with no actual body... key deleted, user mad...

**Review by mattlockyer — CHANGES_REQUESTED** [review]
WDYT?

File: contract/src/stage1/function_call.rs (hunk 2 above)
Comment (BenKurrek) [reply contract/src/stage1/function_call.rs]: We could skip the function call, yea - let me do that.

**Review by mattlockyer — APPROVED** [review]

**Pull request merged**